payloadinterceptor.go

```diff
@@ -23,34 +26,55 @@ import (
 type streamingClientInterceptor struct {
 	connect.StreamingClientConn
 
-	receive func(any, connect.StreamingClientConn) error
-	send    func(any, connect.StreamingClientConn) error
-	onClose func()
+	ctx   context.Context //nolint:containedctx
+	state *state
 
+	requestStarted sync.Once
 	mu             sync.Mutex
 	requestClosed  bool
 	responseClosed bool
-	onCloseCalled  bool
+}
+
+func (s *streamingClientInterceptor) init() {
+	s.requestStarted.Do(func() {
+		header := s.RequestHeader()
+		s.ctx = s.state.start(s.ctx, header)
+	})
 }
 
 func (s *streamingClientInterceptor) Receive(msg any) error {
-	return s.receive(msg, s.StreamingClientConn)
+	if err := s.StreamingClientConn.Receive(msg); err != nil {
+		if !errors.Is(err, io.EOF) {
+			s.state.setError(err)
+		}
+		return err
+	}
+	// Must already be initialized by a call to Send. This is because
+	// Receive blocks until the first message.
+	s.state.receive(s.ctx, msg)
+	return nil
 }
 
 func (s *streamingClientInterceptor) Send(msg any) error {
-	return s.send(msg, s.StreamingClientConn)
+	s.init()
+	if err := s.StreamingClientConn.Send(msg); err != nil {
+		if !errors.Is(err, io.EOF) {
+			s.state.setError(err)
+		}
+		return err
+	}
+	s.state.send(s.ctx, msg)
+	return nil
 }
 
 func (s *streamingClientInterceptor) CloseRequest() error {
+	s.init() // Ensure initialized even under error conditions.
 	err := s.StreamingClientConn.CloseRequest()
 	s.mu.Lock()
+	hasClosed := !s.requestClosed && s.responseClosed
```

Comment:
Is this even right? What does it mean to close the response _before_ closing the request? Does that work? What does it do? I would think that `CloseResponse` is what it means to actually **close** the stream, and if the request side hasn't been closed, it would be implicitly closed as part of closing the response.

Basically, "close request" really means **half**-close, and "close response" really means close. So waiting for both to be called feels potentially brittle.